backend/src/v2/driver/driver.go

```diff
@@ -1850,10 +1897,46 @@ func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.A
 				subTaskName = v.ProducerSubtask
 				outputArtifactKey = v.OutputArtifactKey
 			}
-			// If the sub-task is a DAG, reassign currentTask and run
-			// through the loop again.
-			currentTask = tasks[subTaskName]
-			// }
+
+			if currentTask.GetExecution().GetCustomProperties()["iteration_count"] != nil {
+				glog.V(4).Infof("currentTask, %v, is a ParallelFor DAG. Attempting to gather all producer_subtask iterations of %w", currentTask.TaskName(), subTaskName)
+
+				outputs_list := make([]*pipelinespec.RuntimeArtifact, 0)
+				for i := range currentTask.GetExecution().GetCustomProperties()["iteration_count"].GetIntValue() {
+					// Follow the convention set in mlmd.GetExecutionsInDAG for ParallelFor tasks
+					subTaskIterationName := fmt.Sprintf("%s_idx_%v", subTaskName, i)
```

Comment:
Consider adding a function that abstracts deriving the task name when iteration count is set. This `fmt.Sprintf` is used a lot in this PR and the comment about fllowing the convention of `GetExecutionsInDAG` isn't always present.